### PR TITLE
[LANG] Flutter

### LINKED ISSUE #439: Add Dart/Flutter Support Please!!!
the LspInstall dose not provide any option for dart. Cant really understand what to do.

### PR DESCRIPTION
**Pull request by UltimateOmega**
# Description
This is a new PR based on #681 and #799
Fixes #439 

## How Has This Been Tested?
This PR needs to be tested.

- [ ] The check for Flutter_tools
- [ ]  Custom sdk_path

### CODE REVIEW

**Comment by UltimateOmega** [comment]
@andremeireles @pxsanghyo
Can you guys check this PR?

File: lua/core/which-key.lua [hunk 1]
```diff
@@ -174,6 +174,26 @@ M.setup = function()
   local mappings = O.plugin.which_key.mappings
   local vmappings = O.plugin.which_key.vmappings
 
+  -- Flutter-Tools
```
Comment (ChristianChiarulli) [inline lua/core/which-key.lua:177]: should be included in lang file

File: lua/lang/dart.lua [hunk 2]
```diff
@@ -37,6 +40,10 @@ M.lsp = function()
     return
   end
 
+  if O.lang.dart.flutter_tools.active then
```
Comment (ChristianChiarulli) [inline lua/lang/dart.lua:43]: how does the flutter lang server start?

Comment (ChristianChiarulli) [reply lua/lang/dart.lua:43]: for instance we start separate ones for rust-tools and java-tools

**Review by ChristianChiarulli — CHANGES_REQUESTED** [review]

**Comment by andremeireles** [comment]
I have tested and everything is working except the intellisense (Shift + K), like described by [@pxsanghyo](https://github.com/ChristianChiarulli/LunarVim/pull/799#issuecomment-877751352) in my pull request.

In the file dart.lua, if you remove the lines 43 -46, intellisense comes back.

If the user has standart sdk_path, and those lines are removed, everything will be fine BUT if the user has a custom sdk_path, he/she can't scroll with J and K, up and down unless 

`O.lang.dart.flutter_tools.active = true`

and

`O.lang.dart.sdk_path="/home/custom_path/flutter/bin/cache/dart-sdk/bin/snapshots/analysis_server.dart.snapshot"`

are set.

File: lua/core/which-key.lua (hunk 1 above)
Comment (UltimateOmega) [reply lua/core/which-key.lua]: @ChristianChiarulli  done.

**Comment by UltimateOmega** [comment]
@andremeireles Right now I removed those lines.

File: lua/lang/dart.lua (hunk 2 above)
Comment (UltimateOmega) [reply lua/lang/dart.lua]: rust-tools draws inspiration from akinsho/flutter-tools.nvim
"You can override any options available in the lspconfig setup, this call essentially wraps it and adds some extra flutter specific handlers and utilisation options."
https://github.com/akinsho/flutter-tools.nvim/blob/main/lua/flutter-tools/lsp.lua

File: lua/lang/dart.lua [hunk 3]
```diff
@@ -2,12 +2,32 @@ local M = {}
...
+    flutter_tools = {
+      active = false,
+    },
     sdk_path = "/usr/lib/dart/bin/snapshots/analysis_server.dart.snapshot",
     formatter = {
       exe = "dart",
       args = { "format" },
     },
   }
+  if O.lang.dart.flutter_tools.active then
```
Comment (andremeireles) [inline lua/lang/dart.lua:14]: After moving the shortcut for this file, the shortcut F doesn't work anymore.

Comment (UltimateOmega) [reply lua/lang/dart.lua:14]: @ChristianChiarulli any idea?

Comment (UltimateOmega) [reply lua/lang/dart.lua:14]: if I comment "if O.lang.dart.flutter_tools.active then" it would work again.

Comment (UltimateOmega) [reply lua/lang/dart.lua:14]: but we need the check...

File: lua/lang/dart.lua [hunk 4]
```diff
@@ -2,12 +2,32 @@ local M = {}
...
+    O.plugin.which_key.mappings["F"] = {
```
Comment (abzcoding) [inline lua/lang/dart.lua:15]: @UltimateOmega  change this to line and check it fixes the keybinding not working issue
```lua
O.user_which_key = {
   F = {
     name = "+Flutter",
     c = { ":FlutterCopyProfilerUrl<CR>", "Copy Profile Url" },
     -- ...
   }
}
```

Comment (UltimateOmega) [reply lua/lang/dart.lua]: done!

Comment (UltimateOmega) [reply lua/lang/dart.lua]: but it's now showing up on every file type.

Comment (UltimateOmega) [reply lua/lang/dart.lua]: even when
O.lang.dart.flutter_tools.active = false

Comment (UltimateOmega) [reply lua/lang/dart.lua]: man i hate stylua...

**Comment by iapicca** [comment]
maybe this can be useful https://github.com/akinsho/flutter-tools.nvim

**Comment by UltimateOmega** [comment]
I think I will wait for new_contract to get merged after that I will open a new PR for Flutter

**Comment by UltimateOmega** [comment]
rust tools has been removed, I will show how to install it on the website in the future
_Originally posted by @ChristianChiarulli in https://github.com/ChristianChiarulli/LunarVim/issues/1077#issuecomment-886131993_
I will wait for Chris to show the install for rust first.

**Comment by wwwdata** [comment]
Hi, any updates about Flutter Support? I just started trying out Lunarvim and so far I think it's very amazing. I also disabled the standard lsp for dart and manually started `flutter-tools.nvim` but for me the `Shift-K` key for type information is then also not working.
@UltimateOmega  are you planning to work on Flutter support again?

**Comment by wwwdata** [comment]
By the way, the additional key mappings like `Shift-K` and also `gd` and so on were not registered because this function here was obviously also not called when using the flutter tools plugin to set up LSP.. My current workaround was copy-pasting it and manually doing the same key bindings in the `on_attach` callback

https://github.com/LunarVim/LunarVim/blob/9ece2e5369de46962422837be3dce3b8e889805d/lua/lsp/init.lua#L36-L56

**Comment by pxsanghyo** [comment]
@wwwdata for me `Shift-K` and `gd` works fine.
This is my config:

```lua
lvim.lsp.override = { "dart" }

lvim.plugins = {
  {
    "akinsho/flutter-tools.nvim",
    requires = "nvim-lua/plenary.nvim",
    config = function()
      require("flutter-tools").setup {
        lsp = {
          on_attach = require("lsp").common_on_attach
        }
      }
    end,
  },
}
```

**Pull request closed without merging**